build_modules/lib/src/meta_module.dart

```diff
@@ -167,68 +151,86 @@ Set<AssetId> _localTransitiveDeps(
 /// Creates a map of modules to the entrypoint modules that transitively
 /// depend on those modules.
 Map<AssetId, Set<AssetId>> _findReverseEntrypointDeps(
-    Set<AssetId> entrypointModules, Map<AssetId, Module> modulesById) {
+    Iterable<Module> entrypointModules, Iterable<Module> modules) {
   var reverseDeps = <AssetId, Set<AssetId>>{};
   var assetsToModules = <AssetId, Module>{};
-  for (var module in modulesById.values) {
+  for (var module in modules) {
     for (var assetId in module.sources) {
       assetsToModules[assetId] = module;
     }
   }
-  for (var id in entrypointModules) {
-    for (var moduleDep
-        in _localTransitiveDeps(modulesById[id], assetsToModules)) {
-      reverseDeps.putIfAbsent(moduleDep, () => new Set<AssetId>()).add(id);
+  for (var module in entrypointModules) {
+    for (var moduleDep in _localTransitiveDeps(module, assetsToModules)) {
+      reverseDeps
+          .putIfAbsent(moduleDep, () => new Set<AssetId>())
+          .add(module.primarySource);
     }
   }
   return reverseDeps;
 }
 
-/// Merges [originalModulesById] into a minimum set of [Module]s using the
+/// Merges [modules] into a minimum set of [Module]s using the
 /// following rules:
 ///
-///   * If it is an entrypoint module, skip it.
-///   * Else merge it into a module whose name is a the entrypoint
-///   that import it (create that module if it doesn't exist).
-List<Module> _mergeModules(
-    Map<AssetId, Module> originalModulesById, Set<AssetId> entrypoints) {
-  var modulesById = new Map<AssetId, Module>.from(originalModulesById);
+///   * If it is an entrypoint module do not merge it.
+///   * If it is not depended on my any entrypoint do not merge it.
+///   * If it is depended on by on entrypoint merge it into the entrypoint
+///     modules
+///   * Else merge it into with others that are depended on by the same set of
+///     entrypoints
```

Comment:
Dart highlighters act funny if you indent past the bullet point. Keep this under the bullet.

Reply:
Done